### PR TITLE
Add snap_borders plugin

### PR DESCRIPTION
**Pull request by sloriot**

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h [hunk 1]
```diff
@@ -354,8 +354,9 @@ void reverse_face_orientations(const FaceRange& face_range, PolygonMesh& pmesh)
 
 /**
 * \ingroup PMP_orientation_grp
-* makes each connected component of a closed triangulated surface mesh
-* inward or outward oriented.
+* makes each closed connected component of a triangulated surface mesh
+* inward or outward oriented. If a connected component is not closed,
+* the fact that its orientation is changed or not is not guaranteed.
```
Comment (afabri) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h:359]: "the fact ..."   -->    "the orientation may or may not be changed."

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.5-Ic-42

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Snapping/snap.h [hunk 2]
```diff
@@ -697,6 +727,32 @@ std::size_t split_edges(EdgesToSplitContainer& edges_to_split,
...
+       */
+      auto p1p3 = CGAL::cross_product(p2-p1,p3-p2) * CGAL::cross_product(p0-p3,p1-p0);
+      auto p0p2 = CGAL::cross_product(p1-p0,p1-p2) * CGAL::cross_product(p3-p2,p3-p0);
+      bool first_split_face = (p0p2>p1p3);
+
+      bool is_deg = first_split_face
+                  ? collinear(p0,p1,p2) && collinear(p0,p3,p2)
+                  : collinear(p0,p1,p3) && collinear(p1,p2,p3);
+
+      if (is_deg)
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Snapping/snap.h:753]: Like at other places, I'm more in favor of doing the snap and treating the degenerate faces (it's a split of a border halfedge, it should be rather straightforward?)

**Comment by lrineau** [comment]
This PR now has a conflict.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Snapping/snap.h [hunk 3]
```diff
@@ -151,13 +151,38 @@ void simplify_range(HalfedgeRange& halfedge_range,
             new_tolerance += CGAL::approximate_sqrt(CGAL::squared_distance(new_p, pt));
         }
 
+        if (!CGAL::Euler::does_satisfy_link_condition(edge(h, tm), tm))
+          continue;
+
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Snapping/snap.h:156]: // check that the collapse does not create a new degenerate face

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Snapping/snap.h [hunk 4]
```diff
@@ -151,13 +151,38 @@ void simplify_range(HalfedgeRange& halfedge_range,
...
+        for (halfedge_descriptor he : halfedges_around_target(h, tm))
+          if (he!=h && get(vpm, source(he, tm))==new_p)
+          {
+            call_continue = true;
+            break;
+          }
+        if (call_continue)
+          continue;
+        for (halfedge_descriptor he : halfedges_around_target(opposite(h,tm), tm))
+          if (he!=opposite(h,tm) && get(vpm,source(he, tm))==new_p)
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Snapping/snap.h:167]: should be a collinearity test to handle the cap case

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Snapping/snap.h [hunk 5]
```diff
@@ -151,13 +151,38 @@ void simplify_range(HalfedgeRange& halfedge_range,
...
+          edges_to_test.erase( opoh );
         vertex_descriptor v = Euler::collapse_edge(edge(h, tm), tm);
+
         put(vpm, v, new_p);
         put(tolerance_map, v, new_tolerance);
 
         if(get(range_halfedges, prev_h))
           edges_to_test.insert(prev_h);
-        if(get(range_halfedges, next_h))
+        if(next_h!=opoh && get(range_halfedges, next_h))
```
[the remark this answers is not in the document]
Comment (sloriot) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Snapping/snap.h:185]: avoid inserting a deleted edge in `edges_to_test`

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h (hunk 1 above)
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orientation.h:359]:
```suggestion
* the orientation may or may not be changed or not is not guaranteed.
```

**Pull request merged**